File: src/nexus/backends/storage/remote.py
"""RemoteBackend — ObjectStoreABC proxy for federation content reads.

Proxies content operations to a remote Nexus peer over gRPC via ``RPCTransport``.
Implements the ``ObjectStoreABC`` interface so the kernel can run its natural
VFS pipeline for cross-zone content fetches.

Used by DriverLifecycleCoordinator.resolve_backend() for federation reads
when a local backend miss falls back to a remote origin node. NOT used for
REMOTE deployment profile — that is handled by Rust RemoteBackend installed
via ``sys_setattr(backend_type="remote")``.

Content deletion (``delete_content``) is a deliberate no-op: the kernel
always follows with ``metastore.delete(path)`` which triggers the
server-side delete pipeline.

Issue #844: Converge RemoteNexusFS → NexusFS(profile=REMOTE).
Issue #1133: Unified gRPC transport.
Issue #1134: REMOTE profile migrated to Rust kernel.
"""

from __future__ import annotations

import logging
from typing import TYPE_CHECKING, Any

from nexus.core.object_store import ObjectStoreABC, WriteResult
from nexus.remote.base_client import BaseRemoteNexusFS

if TYPE_CHECKING:
    from nexus.contracts.types import OperationContext
    from nexus.remote.rpc_transport import RPCTransport

logger = logging.getLogger(__name__)


class RemoteBackend(ObjectStoreABC):
    """ObjectStoreABC implementation that proxies to a remote Nexus server.

    Uses ``RPCTransport`` (gRPC) for all RPC calls, with automatic retry
    (tenacity: 3 attempts, exponential backoff 1–10 s) handled by the
    transport layer.

    Args:
        transport: Shared ``RPCTransport`` instance (gRPC channel).
    """

    # RemoteBackend dispatch is handled via DT_EXTERNAL_STORAGE entry_type
    # in metastore, not via capabilities. The server knows the actual
    # backend type and handles reads correctly.
    capabilities: frozenset[str] = frozenset()

    def __init__(self, transport: "RPCTransport") -> None:
        super().__init__()
        self._transport = transport

        # Reuse BaseRemoteNexusFS error handling (static method access)
        self._error_handler = BaseRemoteNexusFS()

    # === Identity ===

    @property
    def name(self) -> str:
        return "remote"

    @property
    def has_root_path(self) -> bool:
        """Remote server always has a configured root path."""
        return True

    # === RPC Transport ===

    def _call_rpc(
        self,
        method: str,
        params: dict[str, Any] | None = None,
        read_timeout: float | None = None,
    ) -> Any:
        """Delegate RPC call to the shared transport."""
        return self._transport.call_rpc(method, params, read_timeout=read_timeout)

    # === Path Resolution ===

    @staticmethod
    def _to_server_path(context: "OperationContext | None") -> str:
        """Extract server-absolute path from OperationContext.

        The kernel sets ``virtual_path`` (the full absolute nexus path) and
        ``backend_path`` (mount-stripped relative path) on the context before
        calling backend methods.  We prefer ``virtual_path`` because it is
        already absolute; fall back to ``backend_path`` with ``/`` prepended.
        """
        if context is not None:
            if context.virtual_path:
                return context.virtual_path
            if context.backend_path:
                bp = context.backend_path
                return bp if bp.startswith("/") else "/" + bp
        return "/"

    # === CAS Content Operations ===

    def write_content(
        self,
        content: bytes,
        content_id: str = "",
        *,
        offset: int = 0,
        context: OperationContext | None = None,
    ) -> WriteResult:
        path = self._to_server_path(context)
        result = self._transport.write_file(path, content)  # Typed RPC — raw bytes
        etag = result.get("etag", "")
        return WriteResult(
            content_id=etag,
            version=etag,
            size=result.get("size", len(content)),
        )

    def read_content(self, content_id: str, context: OperationContext | None = None) -> bytes:
        path = self._to_server_path(context)
        return self._transport.read_file(path, content_id=content_id)

    def delete_content(self, content_id: str, context: OperationContext | None = None) -> None:
        """No-op: server-side deletion is handled by RemoteMetastore.delete().

        The kernel always calls ``metastore.delete(path)`` after
        ``backend.delete_content()``.  In REMOTE mode, ``RemoteMetastore.delete``
        sends the ``delete`` RPC to the server which runs the full delete
        pipeline (CAS content + metadata).  Doing it here as well would be
        redundant and risks a double-delete race.
        """

    def get_content_size(self, content_id: str, context: OperationContext | None = None) -> int:
        path = self._to_server_path(context)
        result = self._call_rpc("sys_stat", {"path": path})
        size: int = int(result.get("size", 0)) if isinstance(result, dict) else 0
        return size

    # === Directory Operations ===

    def mkdir(
        self,
        path: str,
        parents: bool = False,
        exist_ok: bool = False,
        context: OperationContext | None = None,
    ) -> None:
        abs_path = path if path.startswith("/") else "/" + path
        self._call_rpc("mkdir", {"path": abs_path, "parents": parents, "exist_ok": exist_ok})

    def rmdir(
        self,
        path: str,
        recursive: bool = False,
        context: OperationContext | None = None,
    ) -> None:
        abs_path = path if path.startswith("/") else "/" + path
        self._call_rpc("sys_rmdir", {"path": abs_path, "recursive": recursive})

    # === Query Operations ===

    def content_exists(self, content_id: str, context: OperationContext | None = None) -> bool:
        path = self._to_server_path(context)
        result = self._call_rpc("access", {"path": path})
        if isinstance(result, dict):
            return bool(result.get("exists", False))
        return bool(result)

    def list_dir(self, path: str, context: OperationContext | None = None) -> list[str]:
        """List directory contents on the remote server."""
        abs_path = path if path.startswith("/") else "/" + path
        result = self._call_rpc("sys_readdir", {"path": abs_path})
        if isinstance(result, list):
            return [str(item) for item in result]
        if isinstance(result, dict):
            # The RPC handler returns a ``{"files": [...]}`` envelope
            # (``handle_list`` in ``server/rpc/handlers/filesystem.py``),
            # but earlier RemoteBackend code only checked for ``"items"``.
            # Accept both so list operations round-trip correctly.
            items: list[Any] | None = None
            if "files" in result:
                items = result["files"]
            elif "items" in result:
                items = result["items"]
            if items is not None:
                return [
                    str(item.get("path", item.get("name", "")))
                    if isinstance(item, dict)
                    else str(item)
                    for item in items
                ]
        return []

    # === Lifecycle ===

    def close(self) -> None:
        """No-op — transport lifecycle managed by factory."""

File: rust/backends/src/storage/remote.rs
//! Remote ObjectStore — `ObjectStore` trait impl via tonic gRPC.
//!
//! Replaces Python `backends/storage/remote.py`. Content ops use typed
//! Read/Write/Delete RPCs (raw bytes, no base64). Directory ops and
//! stat fall back to the generic Call RPC.
//!
//! Issue #1134: Rust-first connector routing + REMOTE profile.

use std::sync::Arc;

use kernel::abc::object_store::{ObjectStore, StorageError, WriteResult};
use kernel::rpc_transport::RpcTransport;

/// ObjectStore backed by a remote Nexus server via gRPC.
///
/// Server-side NexusFS is the SSOT — this backend is a thin proxy.
/// Path resolution: `backend_path` from kernel routing is the
/// server-relative path; for REMOTE profile the root mount maps
/// backend_path == user_path.
pub(crate) struct RemoteBackend {
    transport: Arc<RpcTransport>,
}

impl RemoteBackend {
    pub fn new(transport: Arc<RpcTransport>) -> Self {
        Self { transport }
    }
}

/// Derive server path from the kernel-provided `backend_path`.
///
/// For REMOTE profile root mount, backend_path == user_path (e.g.
/// "/docs/readme.md"). Ensure leading slash.
fn to_server_path(backend_path: &str) -> String {
    if backend_path.is_empty() {
        "/".to_string()
    } else if backend_path.starts_with('/') {
        backend_path.to_string()
    } else {
        format!("/{backend_path}")
    }
}

impl ObjectStore for RemoteBackend {
    fn name(&self) -> &str {
        "remote"
    }

    fn read_content(
        &self,
        content_id: &str,
        backend_path: &str,
        _ctx: &kernel::kernel::OperationContext,
    ) -> Result<Vec<u8>, StorageError> {
        let path = to_server_path(backend_path);
        self.transport
            .read(&path, content_id)
            .map(|r| r.content)
            .map_err(|e| StorageError::IOError(std::io::Error::other(e)))
    }

    fn write_content(
        &self,
        content: &[u8],
        content_id: &str,
        _ctx: &kernel::kernel::OperationContext,
        _offset: u64,
    ) -> Result<WriteResult, StorageError> {
        let path = to_server_path(content_id);
        let result = self
            .transport
            .write(&path, content, "")
            .map_err(|e| StorageError::IOError(std::io::Error::other(e)))?;
        Ok(WriteResult {
            content_id: result.etag.clone(),
            version: result.etag,
            size: result.size,
        })
    }

    fn delete_content(&self, _content_id: &str) -> Result<(), StorageError> {
        // Server handles content deletion via metastore delete (sys_unlink).
        // Content deletion by hash is not meaningful for remote backends.
        Ok(())
    }

    fn get_content_size(&self, content_id: &str) -> Result<u64, StorageError> {
        // Use sys_stat to get size — content_id is the path for remote.
        let payload = serde_json::json!({ "path": content_id });
        let bytes = serde_json::to_vec(&payload)
            .map_err(|e| StorageError::IOError(std::io::Error::other(e.to_string())))?;
        let (resp, is_error) = self
            .transport
            .call("sys_stat", &bytes)
            .map_err(|e| StorageError::IOError(std::io::Error::other(e)))?;
        if is_error {
            return Err(StorageError::NotFound(content_id.to_string()));
        }
        let value: serde_json::Value = serde_json::from_slice(&resp)
            .map_err(|e| StorageError::IOError(std::io::Error::other(e.to_string())))?;
        value
            .get("size")
            .and_then(|v| v.as_u64())
            .ok_or_else(|| StorageError::NotFound(content_id.to_string()))
    }

    fn mkdir(&self, path: &str, parents: bool, exist_ok: bool) -> Result<(), StorageError> {
        let payload = serde_json::json!({
            "path": path,
            "parents": parents,
            "exist_ok": exist_ok,
        });
        let bytes = serde_json::to_vec(&payload)
            .map_err(|e| StorageError::IOError(std::io::Error::other(e.to_string())))?;
        let (_resp, is_error) = self
            .transport
            .call("mkdir", &bytes)
            .map_err(|e| StorageError::IOError(std::io::Error::other(e)))?;
        if is_error {
            return Err(StorageError::IOError(std::io::Error::other(format!(
                "mkdir failed: {path}"
            ))));
        }
        Ok(())
    }

    fn rmdir(&self, path: &str, recursive: bool) -> Result<(), StorageError> {
        let payload = serde_json::json!({
            "path": path,
            "recursive": recursive,
        });
        let bytes = serde_json::to_vec(&payload)
            .map_err(|e| StorageError::IOError(std::io::Error::other(e.to_string())))?;
        let (_resp, is_error) = self
            .transport
            .call("sys_rmdir", &bytes)
            .map_err(|e| StorageError::IOError(std::io::Error::other(e)))?;
        if is_error {
            return Err(StorageError::IOError(std::io::Error::other(format!(
                "rmdir failed: {path}"
            ))));
        }
        Ok(())
    }

    fn delete_file(&self, path: &str) -> Result<(), StorageError> {
        self.transport
            .delete(path, false)
            .map_err(|e| StorageError::IOError(std::io::Error::other(e)))?;
        Ok(())
    }

    fn rename(&self, old_path: &str, new_path: &str) -> Result<(), StorageError> {
        let payload = serde_json::json!({
            "old_path": old_path,
            "new_path": new_path,
        });
        let bytes = serde_json::to_vec(&payload)
            .map_err(|e| StorageError::IOError(std::io::Error::other(e.to_string())))?;
        let (_resp, is_error) = self
            .transport
            .call("sys_rename", &bytes)
            .map_err(|e| StorageError::IOError(std::io::Error::other(e)))?;
        if is_error {
            return Err(StorageError::IOError(std::io::Error::other(format!(
                "rename failed: {old_path} -> {new_path}"
            ))));
        }
        Ok(())
    }
}
